### PR TITLE
Update persisted queries docs

### PR DESCRIPTION
**Pull request by Meschreiber**
This updates the APQ docs to include both APQs and persisted queries.

### CODE REVIEW

**Comment by BobaFetters** [comment]
Hmmm the roadmap and version checker stuff shouldn't be showing as new changes here since they are already in the main branch, going to see what I can do about fixing that

File: docs/source/fetching/persisted-queries.mdx
Comment (BobaFetters) [inline docs/source/fetching/persisted-queries.mdx:99]: I don't think this section should be part of this doc since it doesn't really pertain to the iOS portion. At most I think if we want to mention/reference the PQL we should link out to a separate docs page for that.

Comment (calvincestari) [reply docs/source/fetching/persisted-queries.mdx:99]: Hmm, I think it's preferable to have one place where the publish instructions are referenced for ease of maintenance but it's a pretty small section so I'm not completely opposed to having it here.

File: docs/source/fetching/persisted-queries.mdx
Comment (calvincestari) [inline docs/source/fetching/persisted-queries.mdx:13]: > If the Apollo iOS client receives this error, it automatically retries the operation with the full operation string.

I don't remember what the default for APQ behaviour is in Apollo iOS, but this may not be true if it's disabled by default.

**Comment by Meschreiber** [comment]
> Hmmm the roadmap and version checker stuff shouldn't be showing as new changes here since they are already in the main branch, going to see what I can do about fixing that

How embarrassing 🙈, thanks for fixing @BobaFetters !

File: docs/source/fetching/persisted-queries.mdx
Comment (Meschreiber) [reply docs/source/fetching/persisted-queries.mdx:13]: Ah, good to know. @BobaFetters can you confirm the default behavior?

Will reword if the retries are disabled by default.

File: docs/source/fetching/persisted-queries.mdx
Comment (Meschreiber) [reply docs/source/fetching/persisted-queries.mdx:99]: I wrestled with this too, since this isn't iOS specific. At the same time, if I was a client maintainer, I'd prefer to have everything in one place, which is why I decided to put it in. If others feel strongly about removing it, we can do that.

Comment (Meschreiber) [reply docs/source/fetching/persisted-queries.mdx:99]: > Is this being duplicated into the Kotlin docs too?
Yes! I'd like to have parity between all client docs on this.

BTW, it'll also be in the ["main PQ" docs](https://github.com/apollographql/docs/pull/496) regardless, just want to make it easier for a client dev by having everything here. Am considering making reusable stubs for any sections that will be duplicated.

File: docs/source/fetching/persisted-queries.mdx
Comment (BobaFetters) [reply docs/source/fetching/persisted-queries.mdx:13]: From what I see in the code it looks like if you have `autoPersistQueries` turned on in the networking then it will automatically retry the request if it fails if the error message contains `PersistedQueryNotFound`, and some other checks like to make sure it hasn't retried already etc

**Review by BobaFetters — APPROVED** [review]
I think this looks good, thanks for this!

File: docs/source/fetching/persisted-queries.mdx
Comment (Meschreiber) [reply docs/source/fetching/persisted-queries.mdx:13]: Thanks for digging! Is this configurable at all?

Comment (BobaFetters) [reply docs/source/fetching/persisted-queries.mdx:13]: Doesn't appear to be, looks like if you turn on auto persisted queries the retry based on the `PersistedQueryNotFound` error will always be how this is handled.

**Comment by Meschreiber** [comment]
JFYI, waiting to merge this until:
1. The ["main" PQ docs](https://github.com/apollographql/docs/pull/496) have been merged.
2. There's parity in the other client docs.

File: docs/source/fetching/persisted-queries.mdx
Comment (AnthonyMDev) [inline docs/source/fetching/persisted-queries.mdx:43]: Before making this feature stable, should we be naming this `legacyAPQs` or just `APQs`? I was under the assumption that once `persistedQueries` was public availability, that the `APQ` feature would be deprecated across the entire platform. But is that assumption correct? @glasser

File: docs/source/fetching/persisted-queries.mdx
Comment (AnthonyMDev) [inline docs/source/fetching/persisted-queries.mdx:54]: It needs to be stated here that for `APQs` to work, you **MUST** include both the `definition` and `operationId`. For `persistedQueries`, you only need the `operationId`, the `definition` is superflous, since the definition is already persisted.

File: docs/source/fetching/persisted-queries.mdx
Comment (AnthonyMDev) [reply docs/source/fetching/persisted-queries.mdx:99]: I'd like to keep this section as well. It would be nice if we could share this block across the different platforms documentation though.

I know we have a folder of `shared` `.md` files that can be placed in-line inside other documentation pages in the iOS repo. But could we share this across the different docs platforms? That way we guarantee we're all using the same verbiage and it can be kept up to date in a single place. @trevorblades Is that possible?

File: docs/source/fetching/persisted-queries.mdx
Comment (AnthonyMDev) [inline docs/source/fetching/persisted-queries.mdx:41]: If you want to generate operation manifest every time you do codegen, this is how you do that, but actually, we recommend that you only generated the manifest as needed using the CLI command as noted below on line 67 of this document.

It looks like we actually need to make an addition to the `generate-operation-manifest` command to make this possible though. So this documentation is accurate to the current state of things. Speaking with @BobaFetters about making the addition so we can update these docs to reflect what we want users to be able to do.

---

You really only need the `operationManifest` for `APQ` if you are also using the safe-listing capabilities or choosing to pre-register your operations with APQs. 

The messaging up above explaining the difference between `APQ` and `persistedQueries` doesn't mention anything about pre-registering documents when using `APQs. It seems to indicate that pre-registering queries is really only a feature of `persistedQueries`.

@glasser Do we still want to even be mentioning the ability to pre-register queries in the APQ world? Or are we trying to remove references to that feature and discourage its use since it was buggy and `persistedQueries` is replacing it?

---

Once we've made the change, I'm wondering if we could make this more clear by organizing this as follows:

It could be 

```markdown
### 1. Generate an operation manifest
{Explain what the operation manifest is for.}
{Explain that the manifest is needed for persisted queries and optional for APQs if you want to pre-register them?}

{Explain the primary way of generating operation manifest, via the CLI command with the `--path` and `--version` arguments}

{Explain the alternative way of adding the `output.operationManifest` to codegen config. Which will re-generate the manifest everytime you run the `generate` command.}

### 2. Configure generated operation models
{Insert the section on `options.operationDocumentFormat` from lines 54-63.}

### 3. Enable persisted queries on `ApolloClient`
{Insert Existing Section}
```

File: docs/source/fetching/persisted-queries.mdx
Comment (AnthonyMDev) [inline docs/source/fetching/persisted-queries.mdx:53]: I'd like a little better explanation of what these options do. Could we add here:

> Setting the `output.operationManifest` option will cause the operation manifest to be generated every time you run code generation.

File: docs/source/fetching/persisted-queries.mdx
Comment (AnthonyMDev) [inline docs/source/fetching/persisted-queries.mdx:110]:
```suggestion
If you want to preregister your manifest as a safelist of trusted operations, you need to publish it to a PQL using the [Rover CLI](/rover/):
```

File: docs/source/fetching/persisted-queries.mdx
Comment (AnthonyMDev) [inline docs/source/fetching/persisted-queries.mdx:117]: `[linked to](#12-link-the-pql-to-variants)`

Looks like this is supposed to link to an anchor in another page that this line was copied from. But this isn't going to be a valid link.

I think this stuff is all the more reason to put this section in a shared block referenced by all the platforms if possible.

**Review by AnthonyMDev — CHANGES_REQUESTED** [review]
Sorry I'm just now getting a chance to take a pass over this!

File: docs/source/fetching/persisted-queries.mdx
Comment (Meschreiber) [reply docs/source/fetching/persisted-queries.mdx:28]: Great feedback! Looks like Router supports APQ, based on this [Slack thread](https://apollograph.slack.com/archives/C04LHRSTLDD/p1690816533569319).

It makes sense you could use APQ on Server without the router, so will clarify that.

File: docs/source/fetching/persisted-queries.mdx
Comment (Meschreiber) [reply docs/source/fetching/persisted-queries.mdx:99]: Good question, @AnthonyMDev ! I didn't think about/realize that these shared `.md` were on per-repo basis.
Maybe easiest just to make a new React component with just text in the main docs repo? Let me know what you think @trevorblades

File: docs/source/fetching/persisted-queries.mdx
Comment (Meschreiber) [reply docs/source/fetching/persisted-queries.mdx:117]: Very true, thanks for catching.

File: docs/source/fetching/persisted-queries.mdx
Comment (AnthonyMDev) [reply docs/source/fetching/persisted-queries.mdx:41]: PR #3172 is up which adds the desired functionality. We should now change these docs as indicated above.

File: docs/source/fetching/persisted-queries.mdx
Comment (martinbonnin) [inline docs/source/fetching/persisted-queries.mdx:33]: Nit: only only PQs require to generate a manifest. APQs work out of the box, at least on Android, there's no codegen configuration to do at all

File: docs/source/fetching/persisted-queries.mdx
Comment (martinbonnin) [reply docs/source/fetching/persisted-queries.mdx:43]: I like `APQ`! They are very easy to setup, we use them in [Confetti](https://github.com/joreilly/Confetti). I would be sad to see them go but I'll follow what's decided here.

File: docs/source/fetching/persisted-queries.mdx
Comment (martinbonnin) [inline docs/source/fetching/persisted-queries.mdx:110]: Any interest in doing that with a Gradle task to avoid downloading an extra tool? Should be relatively easy to do if we want to.

Comment (Meschreiber) [reply docs/source/fetching/persisted-queries.mdx:110]: @AnthonyMDev @BobaFetters , thoughts on this?

Comment (AnthonyMDev) [reply docs/source/fetching/persisted-queries.mdx:110]: Kotlin can do this with a Gradle task, but the iOS build system does not have Gradle or an equivalent tool that is easily available. Using the Rover CLI is the best approach for now on iOS.

File: docs/source/fetching/persisted-queries.mdx
Comment (Meschreiber) [reply docs/source/fetching/persisted-queries.mdx:41]: > Or are we trying to remove references to that feature and discourage its use since it was buggy and persistedQueries is replacing it?

I'd imagine we'd prefer folks to use `persistedQueries` over APQs, so will keep the intro as it is. (I.e. not mention you _can_ preregister queries with APQ.)

Re: removing all mentions of `apq`... not sure. Will bring up with product / PMM.

File: docs/source/fetching/persisted-queries.mdx
Comment (bignimbus) [reply docs/source/fetching/persisted-queries.mdx:43]: @glasser what's your read on this?

File: docs/source/fetching/persisted-queries.mdx
Comment (Meschreiber) [inline docs/source/fetching/persisted-queries.mdx:40]: This is probably because I'm unfamiliar with code generation, but this step label/title confuses me. Is it enough to say "Configure code generation"?

File: docs/source/fetching/persisted-queries.mdx
Comment (Meschreiber) [inline docs/source/fetching/persisted-queries.mdx:53]: I assumed that the `path` and `version` arguments would be used in the same was as they are in the config.json, but looking at the [`generate-operation-manifest`](https://www.apollographql.com/docs/ios/code-generation/codegen-cli#generate-operation-manifest) docs, it looks like `path` is where to read the config from (so you have to update the config anyway?) and the `version` option just spits out the CLI version?

**Comment by Meschreiber** [comment]
Thanks for the thorough review @AnthonyMDev -- I have a couple of follow up questions for you. LMK if you'd prefer to quickly chat.

File: docs/source/fetching/persisted-queries.mdx
Comment (BobaFetters) [reply docs/source/fetching/persisted-queries.mdx:53]: We had some discussions around the CLI command and config json yesterday and there are going to be some changes to it that I am working on today so this documentation will be updated some based on that and hopefully everything will be more clear.

**Comment by BobaFetters** [comment]
Once #3175 is merged I'll take a pass back through these docs to see what might need updated

File: docs/source/fetching/persisted-queries.mdx
Comment (Meschreiber) [reply docs/source/fetching/persisted-queries.mdx:53]: Thank you!

File: docs/source/code-generation/codegen-configuration.mdx
Comment (glasser) [reply docs/source/code-generation/codegen-configuration.mdx:406]: Comma splice?

File: docs/source/fetching/persisted-queries.mdx
Comment (glasser) [inline docs/source/fetching/persisted-queries.mdx:49]: comma splice, and I think "required" is in the wrong place?

File: docs/source/fetching/persisted-queries.mdx
Comment (glasser) [inline docs/source/fetching/persisted-queries.mdx:52]: Should this say subcommand?

File: docs/source/fetching/persisted-queries.mdx
Comment (glasser) [inline docs/source/fetching/persisted-queries.mdx:171]: This isn't about the docs, just the feature — I'm a bit confused as to why iOS can be configured to use GETs for *retries* specifically, vs for all concise requests?

**Review by glasser — APPROVED** [review]
Overall looks pretty good! (I'm skipping over the part where you're planning some CLI changes already.)

File: docs/source/fetching/persisted-queries.mdx
Comment (Meschreiber) [reply docs/source/fetching/persisted-queries.mdx]: https://github.com/apollographql/apollo-ios/pull/3160/commits/99b7a92a8400e606052f4d1e9b225940660e3f27

**Comment by Meschreiber** [comment]
> Once #3175 is merged I'll take a pass back through these docs to see what might need updated

Hey @BobaFetters and @AnthonyMDev, I see [this PR](https://github.com/apollographql/apollo-ios/pull/3175) is out for review.

I'd love to have the updated docs (including the Kotlin version -- cc @martinbonnin) finished by EOD Wednesday. (We're planning on launching PQs Thursday 🚀). Do you think those changes will be merged/documented by then?

File: docs/source/fetching/persisted-queries.mdx
Comment (martinbonnin) [reply docs/source/fetching/persisted-queries.mdx:172]: I think concise requests are always `GET` by default, full requests are always `POST` by default.

So if you want to change your full requests to `GET` you can. Not sure who does this but we have the same configuration option on Apollo Kotlin. I'd be OK leaving it out of the documentation if we think it's not that useful.

File: docs/source/fetching/persisted-queries.mdx
Comment (martinbonnin) [reply docs/source/fetching/persisted-queries.mdx]: > Do we still want to even be mentioning the ability to pre-register queries in the APQ world?

I would vote against mentioning that. I actually don't think you need to pre-register anything with APQ. APQ are completely standalone and don't need pre-registering. This is the beauty of it. On Kotlin, we have an option to generate a manifest in the "old apollo-cli format" (we call it operationOutput) but to my knowledge this isn't used at all by APQ. This was for users who implemented their own persisted queries manually.

File: docs/source/fetching/persisted-queries.mdx
Comment (martinbonnin) [inline docs/source/fetching/persisted-queries.mdx:3]: (absolutely optional) suggestion. Mention the "what" instead of the "how" in the subtitle? 
IMO the subtitle should be about the "what do I want to achieve" to invite the reader to read the "how do I achieve it". 

```suggestion
description: Secure your graph while minimizing request latency
```

File: docs/source/fetching/persisted-queries.mdx
Comment (martinbonnin) [reply docs/source/fetching/persisted-queries.mdx]: My read is that APQs are still a thing and not going anywhere. The "old apollo cli" manifest format on the other hand can be deprecated. I would link the "old apollo cli" manifest format to APQ. 

In Apollo Kotlin we called that format `operationOutput` (I think I can remember the apollo cli using something like that but I wouldn't bet on it). But any format like `legacyManifest` would do.

**Comment by martinbonnin** [comment]
I left a couple comments inline. Also Kotlin PR is up!

* [PR](https://github.com/apollographql/apollo-kotlin/pull/5165)
* [Preview](https://deploy-preview-3160--apollo-ios-docs.netlify.app/fetching/persisted-queries/)

File: docs/source/fetching/persisted-queries.mdx
Comment (Meschreiber) [reply docs/source/fetching/persisted-queries.mdx]: This version is better, thanks @martinbonnin !

**Comment by BobaFetters** [comment]
> > Once #3175 is merged I'll take a pass back through these docs to see what might need updated
> 
> Hey @BobaFetters and @AnthonyMDev, I see [this PR](https://github.com/apollographql/apollo-ios/pull/3175) is out for review.
> 
> I'd love to have the updated docs (including the Kotlin version -- cc @martinbonnin) finished by EOD Wednesday. (We're planning on launching PQs Thursday 🚀). Do you think those changes will be merged/documented by then?

Hoping to get that merged in today once its reviewed and then I can take a look at what doc updates are needed.

File: docs/source/fetching/persisted-queries.mdx
Comment (AnthonyMDev) [reply docs/source/fetching/persisted-queries.mdx]: So does that mean that the new manifest format can still be used to persist queries that work with APQs? Has apollo server been updated to support this?

Comment (AnthonyMDev) [reply docs/source/fetching/persisted-queries.mdx]: To be clear, I understand that APQs _generally_ don't need the manifest, since they are automatically-persisted. I'm talking specifically about users that have been using the existing APQ functionality to work as a safelisting mechanism by pre-registering the operations with a manifest.

Comment (glasser) [reply docs/source/fetching/persisted-queries.mdx]: I don't know much about the uses of the older manifest format here — was it for the legacy operation registry/safelisting feature (which was not connected in any way to APQs)? We will certainly be eventually shutting that feature off (it has been undocumented and barred to new users for a while now).

We're not intending to deprecate APQs, which are an easy to set up performance feature (but don't provide any security benefits).

Comment (AnthonyMDev) [reply docs/source/fetching/persisted-queries.mdx]: Okay, so I was under the impression that the legacy safelisting feature was connected to APQs. If it will be eventually shut off, then we can rename this to just `legacy` and move forward. Thanks @glasser

File: docs/source/fetching/persisted-queries.mdx
Comment (trevorblades) [reply docs/source/fetching/persisted-queries.mdx:99]: https://github.com/apollographql/docs/pull/515 exposes all of these shared bits of MDX as components that you can use in your docs' content:

```mdx
- some
- markdown
- content

<ObtainApiKey />
```

If you want to see the list of components that can be used, you can check out the export names in the `src/content/shared/index.js` file in the docs repo.

Comment (Meschreiber) [reply docs/source/fetching/persisted-queries.mdx:99]: Yay, ty!!!! Will replace text sections in client docs with these tomorrow.

File: docs/source/fetching/persisted-queries.mdx
Comment (martinbonnin) [inline docs/source/fetching/persisted-queries.mdx:40]: Nitpick: from a client point of view, there is a single manifest so I would use singular form here. Of course users with multiple schemas would have multiple manifest but I think it's an edge case and we should optimize for the majority.

**Comment by BobaFetters** [comment]
Made some updates based on the changes in #3175 once that is merged in I will rebase that onto this branch as well.

File: docs/source/fetching/persisted-queries.mdx
Comment (AnthonyMDev) [inline docs/source/fetching/persisted-queries.mdx:40]:
```suggestion
1. **For persisted queries only**: Generate and publish an operation manifest
```

File: docs/source/fetching/persisted-queries.mdx
Comment (AnthonyMDev) [inline docs/source/fetching/persisted-queries.mdx:42]: There is no reason generating and publishing the manifest needs to happen first. And I think it would be clearer that it's only necessary for persisted queries if we put it last like this:

```suggestion
To implement these features, follow these steps.

1. Configure generated operation models
2. Enable persisted queries on `ApolloClient`
3. **For persisted queries only**: Generate and publish operation manifests
```

And then move the "Generate Operation Manifests" section to the bottom.

File: docs/source/fetching/persisted-queries.mdx
Comment (AnthonyMDev) [inline docs/source/fetching/persisted-queries.mdx:46]:
```suggestion
This document only describes the steps that need to be taken by the client to send persisted query requests and create a manifest of the client's operations. Persisted queries also require you to create and link a PQL, and to configure your router to receive persisted query requests. For more information on the other configuration aspects of persisted queries, see the [GraphOS persisted queries documentation](/graphos/operations/persisted-queries).
```

File: docs/source/fetching/persisted-queries.mdx
Comment (AnthonyMDev) [reply docs/source/fetching/persisted-queries.mdx:42]: I'm recognizing that this has been added as one of the pieces that we are going to replace with shared components . So if this suggested change is preferred by others, we should make the change there and then use that component.

**Comment by AnthonyMDev** [comment]
@BobaFetters It looks like you accidentally merged changes from your other PR into this one?

**Comment by BobaFetters** [comment]
@AnthonyMDev should be fixed now, for whatever reason sometimes when you rebase changes from main on the forked branch it shows everything as changes, but if you change the destination branch from main to something else and then back to main it resolves the issue

File: docs/source/fetching/persisted-queries.mdx
Comment (AnthonyMDev) [reply docs/source/fetching/persisted-queries.mdx:41]: I like "Configure generated operation models" more!

File: docs/source/fetching/persisted-queries.mdx
Comment (Meschreiber) [reply docs/source/fetching/persisted-queries.mdx]: I think this order makes the most sense,  even if it sticks the PQ-only step in the middle.
1 - Code generation
2 - Publish operation manifest
3 - Client updates 

(It also matches the implementation narrative in the "[main](https://github.com/apollographql/docs/pull/496/files)" PQ docs, though I'm open to changing it if it's not the most user-intuitive.)

If the sole purpose of putting manifest publication last is to make clearer which step is PQ only, I'd say it'd be better to find another way to make this clearer.

**Comment by BobaFetters** [comment]
We pushed [v1.4.0](https://github.com/apollographql/apollo-ios/releases/tag/1.4.0) of the iOS client today, thinking we should update the version of iOS we require/recommend in the docs to be 1.4.0 instead of 1.3.1 since the docs are based off of changes made and release in 1.4.0

File: docs/source/fetching/persisted-queries.mdx
Comment (BobaFetters) [inline docs/source/fetching/persisted-queries.mdx:23]:
```suggestion
- Apollo iOS (v1.4.0+)
```

Comment (Meschreiber) [reply docs/source/fetching/persisted-queries.mdx]: Ah, thanks for calling that out!

**Pull request merged**
